acceptable/_service.py

```diff
@@ -2,30 +2,88 @@
 # GNU Lesser General Public License version 3 (see the file LICENSE).
 
 """acceptable - Programatic API Metadata for Flask apps."""
+from acceptable import _validation
+
+
+class APIMetadata:
+    """Global datastructure for all services.
+
+    Provides a single point to register apis against, so we can easily inspect
+    and verify uniqueness.
+    """
+
+    def __init__(self):
+        self.services = {}
+        self.api_names = set()
+        self.urls = set()
+
+    def register_service(self, name, group):
+        if (name, group) not in self.services:
+            self.services[name, group] = {}
+
+    def register_api(self, name, group, api):
+        assert api.name not in self.api_names, (
+            'API {} is already registered in service {}'.format(
+                api.name, name)
+        )
+        url_key = (api.url, api.methods)
+        assert url_key not in self.urls, (
+            'URL {} {} is already in service {}'.format(
+                '|'.join(api.methods), api.url, name)
+        )
+        self.api_names.add(api.name)
+        self.urls.add((api.url, api.methods))
+        self.services[name, group][api.name] = api
+
+    def bind(self, flask_app, name, group=None):
+        """Bind the service API urls to a flask app."""
+        for name, api in self.services[name, group].items():
+            # only bind APIs that have views associated with them
+            if api.view_fn is None:
+                continue
+            if name not in flask_app.view_functions:
+                flask_app.add_url_rule(
+                    api.url, name, view_func=api.view_fn, **api.options)
+
+    def bind_all(self, flask_app):
+        for name, group in self.services.items():
+            self.bind(flask_app, name, group)
+
+
+METADATA = APIMetadata()
```

Comment:
All-caps feels odd to me for something that isn't a plain-old-data constant.

Reply:
Right will fix.